### PR TITLE
Run topic proxy sessions' event loops on a thread pool.

### PR DESCRIPTION
**Pull request by aforge**
This helps to limit the resources necessary to run a separate goroutine for every multiplexing session's write event loop.

### CODE REVIEW

File: server/threadpool.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
+/******************************************************************************
+ *
+ *  Description :
+ *    A very basic and naive implementation of thread pool.
+ *
+ *****************************************************************************/
+package main
```
Comment (or-else) [inline server/threadpool.go:7]: This is completely generic. Maybe move it to a separate package?

File: server/threadpool.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+// Task represents a work task to be run on the specified thread pool.
+type Task struct {
+	// Closure to execute.
+	work func()
+}
```
Comment (or-else) [inline server/threadpool.go:13]: The struct wrapper looks unnecessary:
```
type Task func()
```

File: server/cluster.go [hunk 3]
```diff
@@ -1060,6 +1110,10 @@ func (sess *Session) clusterWriteLoop(forTopic string) {
 			// In both cases the msg does not need to be forwarded to the proxy.
 
 		case <-sess.detach:
+			log.Println("cluster: detach msg received", sess.sid)
+			return
+		default:
```
Comment (or-else) [inline server/cluster.go:1115]: Could you please check that this is not going to be called in a very tight loop.

File: server/threadpool.go (hunk 1 above)
Comment (aforge) [reply server/threadpool.go]: Moved to `concurrency`.

File: server/threadpool.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
+/******************************************************************************
+ *
+ *  Description :
+ *    A very basic and naive implementation of thread pool.
+ *
+ *****************************************************************************/
+package main
+
+// Task represents a work task to be run on the specified thread pool.
```
[the remark this answers is not in the document]
Comment (aforge) [reply server/threadpool.go]: Done.

File: server/threadpool.go (hunk 2 above)
Comment (aforge) [reply server/threadpool.go]: Done.

File: server/cluster.go (hunk 3 above)
Comment (aforge) [reply server/cluster.go:1116]: Fair enough. Here are some time deltas between successive calls for a session:
19.1µs
1.0314ms
283.1µs
5.8601ms
15.3µs
94.7µs
175µs
1.6739ms
1.7285ms
8.9µs

Let me look more into it.

Comment (or-else) [reply server/cluster.go:1116]: The `default` case in a `select` is run when no other case is ready. When channels are empty, the `for` will call `default` with no delay burning the CPU cycles.

Comment (or-else) [reply server/cluster.go:1116]: It should be either a ticker or no default.

Comment (aforge) [reply server/cluster.go:1116]: Well, yes. But there's a return statement on line 1118.
So the for loop stops once there are no more ready channels to read from and we exit the function.
What I'm more concerned about is when multiple `clusterWriteLoop` calls are scheduled in a rapid succession (hence, time deltas). In such a case, we'll have to deal with extra scheduling overhead.

Comment (or-else) [reply server/cluster.go:1116]: Why does it take so much time?
I've got a question on architecture.Originally we had a routine per topic-node. Say, 2K topics at 20 nodes gave 40K routines at each node, right? And supposedly it was a problem? 
Right now it's 20 routines shared between all nodes. What if we pin the go routines either to the number of topics or number of nodes * X?

**Review by or-else — APPROVED** [review]

File: server/cluster.go (hunk 3 above)
Comment (aforge) [reply server/cluster.go:1116]: Sure. I'm planning to make the number of goroutines in the pool configurable.
I'll make it number of nodes * 5.

**Pull request merged**